### PR TITLE
Map Director error codes in Provisioner error responses

### PR DESCRIPTION
**Pull request by franpog859**
**Description**

Changes proposed in this pull request:

- add `machinebox/graphql` library to the Provisioner
- map Director's error codes to Provisioner's response codes

**Related issue(s)**

See the issue #82

### CODE REVIEW

File: components/provisioner/third_party/machinebox/graphql/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+## Installation
+Make sure you have a working Go environment. To install graphql, simply run:
+
+```
+$ go get github.com/machinebox/graphql
+```
+
+## Usage
+
```
Comment (majakurcius) [inline components/provisioner/third_party/machinebox/graphql/README.md:20]: Please add an introductory sentence describing what to do with the code below.

Comment (franpog859) [reply components/provisioner/third_party/machinebox/graphql/README.md:20]: Hey! I don't think I should change it. It's just a copied library from the [machinebox repository](https://github.com/machinebox/graphql). The only change I made is a minor contribution which allows us to use it in our specific way

Comment (majakurcius) [reply components/provisioner/third_party/machinebox/graphql/README.md:20]: I see. In that case, please make sure that the appropriate credits and sources are included (see point 4 in the license). 
I think a note in the form of a comment could also be added in modified parts of code. For example, in `client.go` where we have line 9 (`"github.com/kyma-project/control-plane/components/provisioner/third_party/machinebox/graphql"`), we could add a comment saying something like "The original code comes from `https://github.com/machinebox/graphql`, however, this line was modified." or something like this.

File: components/provisioner/internal/director/directorclient.go [hunk 2]
```diff
@@ -247,10 +248,34 @@ func (cc *directorClient) executeDirectorGraphQLCall(directorQuery string, tenan
...
 
 	return nil
 }
+
+func mapDirectorErrorToProvisionerError(egErr gcli.ExtendedError) apperrors.AppError {
+	errorCodeValue, present := egErr.Extensions()["error_code"]
+	if !present {
+		return apperrors.Internal("Failed to read the error code from the error response. Original error: %v", egErr)
+	}
+	errorCode, ok := errorCodeValue.(float64)
```
Comment (Szymongib) [inline components/provisioner/internal/director/directorclient.go:266]: Questions: why casting the error code to `float64`?

File: components/provisioner/internal/director/directorclient.go [hunk 3]
```diff
@@ -247,10 +248,34 @@ func (cc *directorClient) executeDirectorGraphQLCall(directorQuery string, tenan
...
+	if !ok {
+		return apperrors.Internal("Failed to cast the error code from the error response. Original error: %v", egErr)
+	}
+	switch directorApperrors.ErrorType(errorCode) {
+	case directorApperrors.InternalError, directorApperrors.UnknownError:
+		return apperrors.Internal(egErr.Error())
+	case directorApperrors.Unauthorized, directorApperrors.InsufficientScopes:
+		return apperrors.Forbidden(egErr.Error())
```
Comment (Szymongib) [inline components/provisioner/internal/director/directorclient.go:274]: I am not sure about returning the `Forbidden` status in both cases. It seems like a good fit for `InsufficientScopes` but not sure about the `Unauthorized`.

File: components/provisioner/internal/director/directorclient.go (hunk 2 above)
Comment (franpog859) [reply components/provisioner/internal/director/directorclient.go:266]: I checked the type of incoming error code from Director and it was `float64` even if it looks just like `21` for example. I tried to cast it to `int` anyway but it didn't work at all

File: components/provisioner/internal/director/directorclient.go (hunk 3 above)
Comment (franpog859) [reply components/provisioner/internal/director/directorclient.go:274]: Hm, will the `BadRequest` work better for you? The `Forbidden` one was just my first thought for these two

Comment (Szymongib) [reply components/provisioner/internal/director/directorclient.go:274]: What about some equivalent of 502 or even Internal error? It is not actually user error if Provisioner gets 401/403 from Director but more so some issue with the configuration of Control Plane itself, so I am thinking something like `Bad Gateway` would make the most sense.

File: components/provisioner/internal/director/directorclient.go (hunk 2 above)
Comment (Szymongib) [reply components/provisioner/internal/director/directorclient.go:266]: Thanks for insights. It's probably some GraphQL quirks.

**Pull request merged**